governator-core/src/main/java/com/netflix/governator/internal/PreDestroyMonitor.java

```diff
@@ -0,0 +1,237 @@
+        return new ManagedInstanceScopingVisitor(destroyableInstance, context, action).visitEagerSingleton();
+    }
+    
+    /**
+     * final cleanup of managed instances if any
+     */
+    @Override
+    public synchronized void close() throws Exception {
+        if (running.compareAndSet(true, false)) {
+            reqQueueExecutor.shutdown();
+            reqQueueExecutor.awaitTermination(2, TimeUnit.SECONDS);
+            if (!reqQueueExecutor.isTerminated()) {
+                reqQueueExecutor.shutdownNow();
+            }
+            synchronized(scopedCleanupActions) {
+                for (Callable<Void> actions : scopedCleanupActions.values()) {
+                    actions.call();
+                }
+                scopedCleanupActions.clear();
+                scopedCleanupActions = Collections.emptyMap();
+            }
+            
+            
+            
+            for (Callable<Void> action : cleanupActions) {
+                action.call();
+            }     
+            cleanupActions.clear();
+            markerReferenceQueue = null;
+            scopeBindings.clear();
+            scopeBindings = Collections.emptyMap();
+        }
+    }
+    
+    /**
+     * visits bindingScope of managed instance to set up an appropriate strategy for cleanup
+     *  
+     */
+    private final class ManagedInstanceScopingVisitor implements BindingScopingVisitor<Boolean> {       
+        private final Object injectee;
+        private final Object context;
+        private final Iterable<LifecycleAction> lifecycleActions;
+
+        private ManagedInstanceScopingVisitor(Object injectee, Object context, Iterable<LifecycleAction> lifecycleActions) {
+            this.injectee = injectee;
+            this.context = context;
+            this.lifecycleActions = lifecycleActions;
+        }
+
+        /*
+         * add a hard-reference ManagedInstanceAction to cleanupActions. Cleanup is triggered only at injector shutdown.
+         * 
+         */
+        @Override
+        public Boolean visitEagerSingleton() {           
+            cleanupActions.addFirst(new ManagedInstanceAction(injectee, lifecycleActions)); // no need for scopedCleanupActions, return the task directly
+            return true;
+        }
+
+        /*
+         * use ScopeCleanupMarker dereferencing strategy to detect scope closure, add new entry to scopedCleanupActions map
+         * 
+         */
+        @Override
+        public Boolean visitScope(Scope scope) {
+            Provider<ScopeCleanupMarker> scopedMarkerProvider = scope.scope(Key.get(ScopeCleanupMarker.class), markerProvider);
+            ManagedInstanceAction instanceAction = new ManagedInstanceAction(injectee, lifecycleActions);
+            ScopeCleanupMarker marker = scopedMarkerProvider.get();
+            Integer markerKey = System.identityHashCode(marker);
+            synchronized(scopedCleanupActions) {
+                if (scopedCleanupActions.containsKey(markerKey)) {
```

Comment:
it's been a major pain to stick with Java 7 for this (all those inner
classes); seems like CI or some other build constraint is not accepting
Java 8.  You're saying Governator is all Java 8 currently?

On Wed, Jul 6, 2016 at 8:56 PM elandau <notifications@github.com> wrote:

> In
> governator-core/src/main/java/com/netflix/governator/internal/PreDestroyMonitor.java
> <https://github.com/Netflix/governator/pull/307#discussion_r69848482>:
>
> > +            cleanupActions.addFirst(new ManagedInstanceAction(injectee, lifecycleActions)); // no need for scopedCleanupActions, return the task directly
> > +            return true;
> > +        }
> > +
> > +        /*
> > +         * use ScopeCleanupMarker dereferencing strategy to detect scope closure, add new entry to scopedCleanupActions map
> > +         *
> > +         */
> > +        @Override
> > +        public Boolean visitScope(Scope scope) {
> > +            Provider<ScopeCleanupMarker> scopedMarkerProvider = scope.scope(Key.get(ScopeCleanupMarker.class), markerProvider);
> > +            ManagedInstanceAction instanceAction = new ManagedInstanceAction(injectee, lifecycleActions);
> > +            ScopeCleanupMarker marker = scopedMarkerProvider.get();
> > +            Integer markerKey = System.identityHashCode(marker);
> > +            synchronized(scopedCleanupActions) {
> > +                if (scopedCleanupActions.containsKey(markerKey)) {
>
> Since we're on Java8 maybe we can start using things like map#compute
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Netflix/governator/pull/307/files/5c39381496a87a3c2cac2fcca569b4f585e99859#r69848482>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACIO-7GVJxvd-GTn-9X6isJte5zolELjks5qTHj0gaJpZM4JGi7C>
> .
>